File: src/leetcode/P127_WordLadder.java
package leetcode;

import java.util.HashSet;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

/*
 * 	Given two words (beginWord and endWord), and a dictionary's word list,
 *  find the length of shortest transformation sequence
 *   from beginWord to endWord, such that:

	Only one letter can be changed at a time
	Each intermediate word must exist in the word list
	For example,
	
	Given:
	beginWord = "hit"
	endWord = "cog"
	wordList = ["hot","dot","dog","lot","log"]
	As one shortest transformation is "hit" -> "hot" -> "dot" -> "dog" -> "cog",
	return its length 5.
	
	Note:
	Return 0 if there is no such transformation sequence.
	All words have the same length.
	All words contain only lowercase alphabetic characters.
 */

/**
 * @author      zxwtry
 * @email       zxwtry@qq.com
 * @project     OJ
 * @package     leetcode
 * @file        P127_WordLadder.java
 * @type        P127_WordLadder
 * @date        2017年5月12日 下午10:28:45
 * @details     AC 103ms 78.74%
 */
public class P127_WordLadder {
    static public class Solution {
        public int ladderLength(String b, String e, List<String> w) {
            int sn = b == null ? 0 : b.length();
            if (sn == 0) return  1;
            HashSet<String> nv = new HashSet<String>(w);
            HashSet<String> hv = new HashSet<String>();
            Queue<String> q = new LinkedList<String>();
            q.add(b);
            nv.add(b);
            char[] cs = new char[sn];
            boolean isFind = false;
            int ans = 1;
            while (! q.isEmpty()) {
                int size = q.size();
                ans ++;
                while (size -- > 0) {
                    String n = q.poll();
                    for (int i = 0; i < sn; i ++) cs[i] = n.charAt(i);
                    for (int i = 0; i < sn; i ++) {
                        for (char c = 'a'; c <= 'z'; c ++) {
                            cs[i] = c;
                            String nn = new String(cs);
                            if (! nv.contains(nn)) continue;
                            if (hv.add(nn)) q.add(nn);
                            isFind |= nn.equals(e);
                        }
                        cs[i] = n.charAt(i);
                    }
                }
                if (isFind) break;
                nv.removeAll(hv);
                hv.clear();
            }
            return isFind ? ans : 0;
        }
    }
}

File: python/P127_WordLadder.py
#coding=utf-8

'''
    url: leetcode.com/problems/word-ladder
    @author:     zxwtry
    @email:      zxwtry@qq.com
    @date:       2017年5月12日
    @details:    Solution:  902ms 44.52%
'''

from jinja2._compat import unichr

class Solution(object):
    def ladderLength(self, b, e, w):
        """
        :type b: str
        :type e: str
        :type w: List[str]
        :rtype: int
        """
        nv = set(w)
        hv = set()
        sn = len(b)
        q = []
        nv.add(b)
        q.append(b)
        isFind = False
        count = 2
        while True:
            l = len(q)
            if l == 0: break
            while l > 0:
                l -= 1
                s = q[0]
                q.remove(q[0])
                cs = [s[i] for i in range(sn)]
                for i in range(sn):
                    for j in range(26):
                        cs[i] = unichr(97+j)
                        ns = "".join(cs)
                        if not ns in nv: continue
                        if not ns in hv:
                            hv.add(ns)
                            q.append(ns)
                        if ns == e: isFind = True
                    cs[i] = s[i]
            if isFind: break
            count += 1
            for v in hv: nv.remove(v)
            hv.clear()
        return 0 if not isFind else count
        

if __name__ == "__main__":
    b, e = "hit", "cog"
    w = ["hot","dot","hog","dog","lot","log","cog"]
#     b, e = "a", "c"
#     w = ["a", "b", "c"]
    print(Solution().ladderLength(b, e, w))
